pkg/controllers/provisioningv2/rke2/bootstrap/controller_test.go

```diff
@@ -0,0 +1,245 @@
+	args := m.Called(namespace, selector)
+	return args.Get(0).([]*v1alpha4.Machine), args.Error(1)
+}
+func (m *machineCacheMock) AddIndexer(indexName string, indexer capicontrollers.MachineIndexer) {}
+func (m *machineCacheMock) GetByIndex(indexName, key string) ([]*v1alpha4.Machine, error) {
+	args := m.Called(indexName, key)
+	return args.Get(0).([]*v1alpha4.Machine), args.Error(1)
+}
+
+func Test_getBootstrapSecret(t *testing.T) {
+	type args struct {
+		secretName             string
+		os                     string
+		token                  string
+		expectedToken          string
+		name                   string
+		installScript          string
+		namespaceName          string
+		path                   string
+		command                string
+		scriptName             string
+		body                   string
+		serviceAccountName     string
+		version                string
+		expectedSecret         string
+		expectedNamespace      string
+		expectedServiceAccount string
+	}
+
+	tests := []struct {
+		name string
+		args args
+	}{
+		{
+			name: "Checking Linux Install Script",
+			args: args{
+				token:              "thisismylinuxtokenandiwillprotectit",
+				version:            "v1.21.5+rke2r2",
+				os:                 "linux",
+				secretName:         "mybestlinuxsecret",
+				command:            "sh",
+				scriptName:         "run.sh",
+				namespaceName:      "myfavoritelinuxnamespace",
+				path:               "/system-agent-install.sh",
+				serviceAccountName: "linux-test-sa",
+				body:               "#!/usr/bin/env sh",
+			},
+		},
+		{
+			name: "Checking Windows Install Script",
+			args: args{
+				token:              "thisismywindowstokenandiwillprotectit",
+				version:            "v1.21.5+rke2r2",
+				os:                 "windows",
+				secretName:         "mybestwindowssecret",
+				namespaceName:      "myfavoritewindowsnamespace",
+				serviceAccountName: "windows-test-sa",
+				body:               "Invoke-WinsInstaller @PSBoundParameters",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// arrange
+			expectHash := sha256.Sum256([]byte("test"))
+			expectEncodedHash := base64.URLEncoding.EncodeToString(expectHash[:])
+			a := assert.New(t)
+
+			handler := handler{
+				serviceAccountCache: getServiceCacheMock(),
+				secretCache:         getSecretCacheMock(),
+				deploymentCache:     getDeploymentCacheMock(),
+				machineCache:        getMachineCacheMock(tt.args.os),
+			}
+
+			//act
+			secret, err := handler.getBootstrapSecret("test", "test", []v1.EnvVar{})
+
+			// assert
+			a.Nil(err)
+			a.NotNil(secret)
+
+			a.Equal("test", secret.Name)
+			a.Equal("test", secret.Namespace)
+			a.Equal("rke.cattle.io/bootstrap", string(secret.Type))
+			data := string(secret.Data["value"])
+			a.Contains(data, "CATTLE_ROLE_NONE=true")
+			a.Contains(data, fmt.Sprintf("CATTLE_TOKEN=\"%s\"", expectEncodedHash))
+			a.Contains(data, tt.args.body)
```

Comment:
Can we add a check here to ensure that the correct script (windows/linux) is used?

Reply:
addressed.